File: 746-min-cost-climbing-stairs/746-min-cost-climbing-stairs.py
class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        dp = [0] * len(cost)
        
        for i in range(len(cost)):
            dp[i] = cost[i] + min(dp[i-1], dp[i-2])
            
        return min(dp[-1], dp[-2])

File: 746-min-cost-climbing-stairs/746-min-cost-climbing-stairs.cpp
class Solution {
public:
    unordered_map<int, int> memo;
    int dp(int idx, vector<int>& cost) {
        if (idx <= 1)
            return cost[idx];
        if (memo[idx]) return memo[idx];
        memo[idx] = cost[idx] + min(dp(idx - 1, cost), dp(idx - 2, cost));
        return memo[idx];
    }
    int minCostClimbingStairs(vector<int>& cost) {
        int n = cost.size();
        return min(dp(n - 1, cost), dp(n - 2, cost));
    }
};